workloads/imagenet/imagenet_pytorch/workload.py

```diff
@@ -0,0 +1,235 @@
+
+  def _build_dataset(self,
+      data_rng: spec.RandomState,
+      split: str,
+      data_dir: str,
+      batch_size: int):
+
+    is_train = (split == "train")
+
+    normalize = transforms.Compose([
+        transforms.ToTensor(),
+        transforms.Normalize(
+            mean=[i/255 for i in self.train_mean], 
+            std=[i/255 for i in self.train_stddev])
+    ])
+    transform_config = {
+        "train": transforms.Compose([
+            transforms.RandomResizedCrop(
+                self.center_crop_size,
+                scale=self.scale_ratio_range,
+                ratio=self.aspect_ratio_range),
+            transforms.RandomHorizontalFlip(),
+            normalize]),
+        "test": transforms.Compose([
+            transforms.Resize(self.resize_size),
+            transforms.CenterCrop(self.center_crop_size),
+            normalize])
+    }
+
+    folder = {'train': 'train', 'test': 'val'}
+
+    dataset = ImageFolder(
+        os.path.join(data_dir, folder[split]),
+        transform=transform_config[split])
+
+    dataloader = torch.utils.data.DataLoader(
+        dataset,
+        batch_size=batch_size,
+        shuffle=is_train,
+        num_workers=5,
+        pin_memory=True,
+        # Note (runame): not sure why we do this?
+        drop_last=is_train,
+        # Note (runame): if we want to use fast_collate,
+        # we need to change some things
+        #collate_fn=fast_collate
+    )
+
+    if is_train:
+      dataloader = cycle(dataloader)
+
+    return dataloader
+
+  def preprocess_for_train(
+      self,
+      selected_raw_input_batch: spec.Tensor,
+      selected_label_batch: spec.Tensor,
+      train_mean: spec.Tensor,
+      train_stddev: spec.Tensor,
+      rng: spec.RandomState) -> spec.Tensor:
+    del train_mean
+    del train_stddev
+    del rng
+    return self.preprocess_for_eval(
+        selected_raw_input_batch, selected_label_batch, None, None)
+
+  def preprocess_for_eval(
+      self,
+      raw_input_batch: spec.Tensor,
+      raw_label_batch: spec.Tensor,
+      train_mean: spec.Tensor,
+      train_stddev: spec.Tensor) -> spec.Tensor:
+    del train_mean
+    del train_stddev
+    return (raw_input_batch.float().to(DEVICE), raw_label_batch.to(DEVICE))
+
+  def init_model_fn(self, rng: spec.RandomState) -> spec.ModelInitState:
+    torch.random.manual_seed(rng[0])
+    model = resnet50()
+    if torch.cuda.device_count() > 1:
+      model = torch.nn.DataParallel(model)
+    model.to(DEVICE)
+    return model, None
+
+  def model_fn(
+      self,
+      params: spec.ParameterContainer,
+      input_batch: spec.Tensor,
+      model_state: spec.ModelAuxiliaryState,
+      mode: spec.ForwardPassMode,
+      rng: spec.RandomState,
+      update_batch_norm: bool) -> Tuple[spec.Tensor, spec.ModelAuxiliaryState]:
+    del model_state
+    del rng
+    del update_batch_norm
+
+    model = params
+
+    if mode == spec.ForwardPassMode.EVAL:
+        model.eval()
```

Comment:
I think we have to be a bit more careful here.

Currently, `update_batch_norm` is ignored and basically replaced by the `mode`.

But this might be not what a submitter would expect. If they would e.g. perform a line-search that "tests" different step sizes before committing to one, it would be possible to want to not update the BatchNorm statistics for those "test" steps. In this case, the network is in `train` mode but you still would want `update_batch_norm=false` (but perhaps you would want to compute the updated BN statistics anyway in case the current point will be accepted?)

Reply:
I have changed the code to allow not updating the batch norm statistics during training. Not sure if there is a more efficient way to do it, since now it loops through all `model.modules()` once before each forward pass in train mode. This is necessary to ensure that the statistics are tracked or not dependent on the `update_batch_norm` argument, since this could change for each forward pass. This overhead is the same for all PyTorch submissions, but we have to see if it introduces a significant overhead compared to the Jax submission.

Reply:
This might be a great question for @mikerabbat :)

Perhaps @znado could also provide his opinion on whether this sounds like the best approach.